Feature: Greenlight3 import

pizkaz opened the pull request:
# Fixes #2664

## Type

- Bugfix
- **Feature**
- Documentation
- Refactoring (e.g. Style updates, Test implementation, etc.)
- Other (please describe):

## Checklist

- [x] Code updated to current develop branch head
- [x] Passes CI checks
- [x] Is a part of an issue
- [x] Tests added for the bugfix or newly implemented feature, describe below why if not
- [x] Changelog is updated
- [x] Documentation of code and features exists

## Changes

- Added a new command `import:greenlight-v3`
- Legacy room access codes can now be alphanumeric (because GL3 uses 6-chars-alphanumeric codes)
- Added / adapted tests for new features and changed behavior (room access codes)

## Other information

**ToDo**:
- [x] GL3 route handling
- [x] Documentation
- [x] e2e Test of a GL3 migration (manually)

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added Greenlight v3 import command for migrating rooms, users, and presentations from Greenlight v3 instances.
  * Support for legacy alphanumeric access codes alongside numeric codes.

* **Improvements**
  * User avatar now displays initials using first two letters of first name when last name is unavailable.
  * Updated migration documentation with Greenlight v3 workflow and instructions.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

pizkaz committed:
feat: Greenlight 3 import command

samuelwei commented on resources/js/components/RoomShareButton.vue:
```diff
@@ -115,8 +115,10 @@ const roomUrl = computed(() => {
-  return String(props.room.access_code)
-    .match(/.{1,3}/g)
-    .join("-");
+  return isNaN(props.room.access_code)
```
That would format 6 digit numeric codes with a dash.
However when entering the 6 digit code there is no dash input mask

samuelwei commented on app/Http/Requests/UpdateRoomSettings.php:
```diff
@@ -42,12 +42,17 @@ public function rules()
-        $rules = ['numeric', 'digits:'.$digits, 'bail'];
+        $rules = $alphanumeric
+               ? ['alpha_num:ascii', 'lowercase', 'size:6', 'bail']
+               : ['numeric', 'digits:'.$digits, 'bail'];
```
```suggestion
        $legacy = $current == $incoming && strlen($current) == 6);

        $rules = $legacy
               ? ['alpha_num:ascii', 'lowercase', 'size:6', 'bail']
               : ['numeric', 'digits:9', 'bail'];
```

pizkaz commented:
I think this is ready to merge.

samuelwei commented on app/Console/Commands/ImportGreenlight3Command.php:
```diff
@@ -0,0 +1,395 @@
+
+        // ask user what room type the imported rooms should get
+        $roomType = $this->option('room-type') != null
+            ? RoomType::firstWhere('name', $this->option('room-type'))->id
```
Is this case sensitive? What happens if the name is invalid (no room type with this name)?

pizkaz replied:
It is not. I added a note to the help message and handled 'No query results' errors.

samuelwei replied:
It's not case sensitive for MySQL/MariaDB but for PostgreSQL:

<img width="954" height="107" alt="image" src="https://github.com/user-attachments/assets/6a2d1fa8-70a4-4a68-8ab0-dfc092f60c7e" />

pizkaz replied:
Consistency. So important!

samuelwei replied:
Maybe use whereLike (https://laravel.com/docs/12.x/queries#additional-where-clauses)

samuelwei commented on app/Console/Commands/ImportGreenlight3Command.php:
```diff
@@ -0,0 +1,395 @@
+            );
+
+        // ask user to add prefix to room names
+        $prefix = $this->option('no-confirm') ? null : text(
```
Why does "do not ask if the import should be committed" effect the prefix ? Maybe add an option to set the prefix ?

pizkaz replied:
I was having a hard time getting `--prefix-path=''` to work as intended. Will try again.

pizkaz replied:
Done

samuelwei commented on app/Console/Commands/ImportGreenlight3Command.php:
```diff
@@ -0,0 +1,395 @@
+            );
+
+        // ask user for presentation path
+        $presentationPath = $this->option('presentation-path')
```
Can we validate if the path exists and throw an error if the path is not a directory ?

pizkaz replied:
Done

samuelwei commented on app/Console/Commands/ImportGreenlight3Command.php:
```diff
@@ -0,0 +1,395 @@
+                DB::commit();
+                $this->info('Import completed');
+            } else {
+                DB::rollBack();
```
What about files that we copied/moved ?

pizkaz replied:
Moved files are now stored in an associative array and moved back to their original places on rollback.

samuelwei commented on app/Console/Commands/ImportGreenlight3Command.php:
```diff
@@ -0,0 +1,395 @@
+                // we have to import it as firstname and ask the users or admins to correct it later if desired
+                $dbUser->firstname = $user->name;
+                $dbUser->lastname = '';
+                $dbUser->password = $user->external_id ? Hash::make(Str::random()) : $user->password_digest;
```
Have you checked if GL 3 still using bcrypt ?

pizkaz replied:
We can log in with a local user after a Greenlight 3 to PILOS migration.

samuelwei commented on app/Console/Commands/ImportGreenlight3Command.php:
```diff
@@ -0,0 +1,395 @@
+        $failed = 0;
+
+        foreach ($roomMap as $gl3RoomId => $pilosRoomId) {
+            $attachments = DB::connection('greenlight')->table('active_storage_attachments')->where('record_id', $gl3RoomId)->get('blob_id');
```
Can be do a join here instead of n + 1 queries ?

pizkaz replied:
Done

samuelwei commented on docs/docs/administration/08-advanced/06-migrate-greenlight.md:
```diff
@@ -26,8 +26,17 @@ ports:
+
 ## Running migration command
 
+The command will output the process of the import and imforms about failed user, room and shared access import.
```
```suggestion
The command will output the process of the import and informs about failed user, room and shared access import.
```

pizkaz replied:
Done

samuelwei commented on docs/docs/administration/08-advanced/06-migrate-greenlight.md:
```diff
@@ -26,8 +26,17 @@ ports:
 Also make sure the internal firewall of the OS and no external firewall is not blocking access to the port and from the host PILOS is running on.
 
+If you want to import Room presentations, copy Greenlight's active storage directory to the PILOS app storage at `/storage/app/migration/presentations` and specify `--presentation-path=migration/presentations` at the command line.
+Successfully imported presentation files will be moved to a different location.
```
This is currently only possible for GL3 import. Move to GL3 section until added to GL2 import

pizkaz replied:
Done

samuelwei commented on docs/docs/administration/08-advanced/06-migrate-greenlight.md:
```diff
@@ -42,9 +51,40 @@ docker compose exec app pilos-cli import:greenlight-v2   {host : ip or hostname
+
+Options:
+      --no-confirm                             do not ask if the import should be committed
+      --default-role[=DEFAULT-ROLE]            default role for imported external users
```
name of the default role ....

pizkaz replied:
Done

samuelwei commented on docs/docs/administration/08-advanced/06-migrate-greenlight.md:
```diff
@@ -42,9 +51,40 @@ docker compose exec app pilos-cli import:greenlight-v2   {host : ip or hostname
+Options:
+      --no-confirm                             do not ask if the import should be committed
+      --default-role[=DEFAULT-ROLE]            default role for imported external users
+      --room-type[=ROOM-TYPE]                  room type for imported rooms
```
name of the room type ....

pizkaz replied:
Done

samuelwei commented on docs/docs/administration/08-advanced/06-migrate-greenlight.md:
```diff
@@ -42,9 +51,40 @@ docker compose exec app pilos-cli import:greenlight-v2   {host : ip or hostname
+      --no-confirm                             do not ask if the import should be committed
+      --default-role[=DEFAULT-ROLE]            default role for imported external users
+      --room-type[=ROOM-TYPE]                  room type for imported rooms
+      --presentation-path[=PRESENTATION-PATH]  path to room presentations
```
relative to /storage/app/

pizkaz replied:
Done

samuelwei commented on app/Console/Commands/ImportGreenlight3Command.php:
```diff
@@ -0,0 +1,430 @@
+    {
+        try {
+            // Ask user what room type the imported rooms should get
+            $roomType = ! is_null($this->option('room-type'))
```
<img width="457" height="415" alt="Image" src="https://github.com/user-attachments/assets/9c3e3f31-1eaa-4806-b3ef-0f263f91365b" />

pizkaz replied:
I want to know if the option was specified, not if it exists in the command signature.

samuelwei replied:
You're right, sorry

samuelwei committed:
Fix case-sensitive issue for command options

samuelwei commented:
@pizkaz What do you think about these global settings?
<img width="1123" height="831" alt="image" src="https://github.com/user-attachments/assets/74dc3f9a-a92b-426d-97b2-9f45acd1148b" />

<img width="1120" height="551" alt="image" src="https://github.com/user-attachments/assets/b9fe1595-e164-4aca-baa3-f341825fbd41" />

samuelwei commented on docs/docs/administration/08-advanced/06-migrate-greenlight.md:
```diff
@@ -42,9 +48,43 @@ docker compose exec app pilos-cli import:greenlight-v2   {host : ip or hostname
+Arguments:
+  host                                         ip or hostname of postgres database server
+  port                                         port of postgres database server
+  database                                     greenlight database name, see greenlight .env variable DB_NAME
```
GL3 .env file doesn't has this field anymore (https://github.com/bigbluebutton/greenlight/blob/master/sample.env#L16)
Same issue in the command signiture

pizkaz replied:
I suggest we strip the comment but leave the option in.

samuelwei replied:
Maybe just: `greenlight database name, see greenlight .env` ?

pizkaz replied:
Or even just `greenlight database name`.

samuelwei commented on app/Console/Commands/ImportGreenlight3Command.php:
```diff
@@ -0,0 +1,430 @@
+                    $presentation = new UploadedFile(Storage::path($path), $blob->filename);
+
+                    // Construct RoomFile
+                    $file = new RoomFile;
+                    $file->path = $path;
+                    $file->filename = $blob->filename;
+                    $file->use_in_meeting = true;
+                    $room = Room::find($pilosRoomId);
+
+                    // Save file and room, delete source file
+                    $presentation->store($room->id);
+                    $room->files()->save($file);
```
```suggestion
                    // Construct RoomFile
                    $room = Room::find($pilosRoomId);
                    $file = new RoomFile;
                    $file->path = $presentation->store($room->id);
                    $file->filename = $blob->filename;
                    $file->use_in_meeting = true;

                    // Save file and room, delete source file
                    $room->files()->save($file);
```

Otherwise path contains the location inside the import path and not the path after moving the file to its final destination

samuelwei committed:
Fix presentation import

samuelwei committed:
Fix user avatar for imported users

samuelwei commented on app/Http/Controllers/api/v1/RoomController.php:
```diff
@@ -467,7 +467,7 @@ public function authenticate(Room $room, RoomAuthRequest $request)
             $accessCode = $request->access_code;
 
-            if (is_numeric($accessCode) && $room->access_code == $accessCode) {
+            if ($room->access_code == $accessCode) {
```
GL3 create lower case access codes only (https://github.com/bigbluebutton/greenlight/blob/master/app/controllers/api/v1/room_settings_controller.rb#L81)
Should we keep access code case-sensitive or make it case-insensitive (frontend or backend)?

pizkaz replied:
Let's stick with case-sensitive matching.

samuelwei commented on app/Models/Room.php:
```diff
@@ -338,7 +338,8 @@ public function getModeratorOnlyMessage()
         if ($this->access_code != null) {
-            $message .= '<br>'.__('rooms.invitation.code').': '.implode('-', str_split($this->access_code, 3));
+            $message .= '<br>'.__('rooms.invitation.code').': ';
+            $message .= $this->legacy_code ? $this->access_code : implode('-', str_split($this->access_code, 3));
```
I doubt this works. legacy_code is not a property to my knowledge

pizkaz replied:
Hmm I believe it was at some point in the evolution of this PR. :thinking: 
Shouldn't there be a failing test for this?

pizkaz replied:
Found it: app/Http/Resources/Room.php line 58

samuelwei replied:
I'm working on a fix and a new test to cover this

pizkaz commented:
> @pizkaz What do you think about these global settings? <img alt="image" width="1123" height="831" src="https://private-user-images.githubusercontent.com/4281791/554155805-74dc3f9a-a92b-426d-97b2-9f45acd1148b.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzE5NDQzODgsIm5iZiI6MTc3MTk0NDA4OCwicGF0aCI6Ii80MjgxNzkxLzU1NDE1NTgwNS03NGRjM2Y5YS1hOTJiLTQyNmQtOTdiMi05ZjQ1YWNkMTE0OGIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDIyNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjAyMjRUMTQ0MTI4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDRjZjFiNDAyMGU5NWNkZWQ0ODM0MTc1NDU0MDc3MzA0NjBhMmI5MWJiNWI3YzQ3ZjU3NmI5NmM2NTk0YTVjYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.O21_xPRIfuV6xCMvYf70xriWB8AG5maYGoZ4ZBqCvj0">
> <img alt="image" width="1120" height="551" src="https://private-user-images.githubusercontent.com/4281791/554155888-b9fe1595-e164-4aca-baa3-f341825fbd41.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzE5NDQzODgsIm5iZiI6MTc3MTk0NDA4OCwicGF0aCI6Ii80MjgxNzkxLzU1NDE1NTg4OC1iOWZlMTU5NS1lMTY0LTRhY2EtYmFhMy1mMzQxODI1ZmJkNDEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDIyNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjAyMjRUMTQ0MTI4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NjljOWE4YWRkYjYzZDBkOTgyNzY5NjUyOTA3YmE4YzU2ZDFlNTVkMTA3M2M3MTM0NDNmNzNlNDU2YjcwNGMyZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.bG7xTHLbw_84yFtlGBeFElyu4AMr_5QunHirB07Sb5E">

Hmm, I'm unsure if the benefit justifies the cost of implementing and testing imports of these settings.

samuelwei commented:
> > @pizkaz What do you think about these global settings? <img alt="image" width="1123" height="831" src="https://private-user-images.githubusercontent.com/4281791/554155805-74dc3f9a-a92b-426d-97b2-9f45acd1148b.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzE5NDQzODgsIm5iZiI6MTc3MTk0NDA4OCwicGF0aCI6Ii80MjgxNzkxLzU1NDE1NTgwNS03NGRjM2Y5YS1hOTJiLTQyNmQtOTdiMi05ZjQ1YWNkMTE0OGIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDIyNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjAyMjRUMTQ0MTI4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDRjZjFiNDAyMGU5NWNkZWQ0ODM0MTc1NDU0MDc3MzA0NjBhMmI5MWJiNWI3YzQ3ZjU3NmI5NmM2NTk0YTVjYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.O21_xPRIfuV6xCMvYf70xriWB8AG5maYGoZ4ZBqCvj0">
> > <img alt="image" width="1120" height="551" src="https://private-user-images.githubusercontent.com/4281791/554155888-b9fe1595-e164-4aca-baa3-f341825fbd41.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzE5NDQzODgsIm5iZiI6MTc3MTk0NDA4OCwicGF0aCI6Ii80MjgxNzkxLzU1NDE1NTg4OC1iOWZlMTU5NS1lMTY0LTRhY2EtYmFhMy1mMzQxODI1ZmJkNDEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDIyNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjAyMjRUMTQ0MTI4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NjljOWE4YWRkYjYzZDBkOTgyNzY5NjUyOTA3YmE4YzU2ZDFlNTVkMTA3M2M3MTM0NDNmNzNlNDU2YjcwNGMyZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.bG7xTHLbw_84yFtlGBeFElyu4AMr_5QunHirB07Sb5E">
> 
> Hmm, I'm unsure if the benefit justifies the cost of implementing and testing imports of these settings.

Yes, let's not do that. The documentation only states: "PILOS provides an easy to use command to import all greenlight users (incl. ldap), rooms and shared accesses."

samuelwei committed:
Add test to check moderator only message

samuelwei committed:
Refactor legacy access code

samuelwei commented on resources/js/components/RoomShareButton.vue:
```diff
@@ -147,8 +147,10 @@ const roomUrl = computed(() => {
 });
 
 const formattedAccessCode = computed(() => {
```
A 6 digit numeric code is still splitted, however the UI for entering the code is not showing dashes.

samuelwei replied:
fixed

samuelwei committed:
Fix legacy code display bug

samuelwei commented on app/Console/Commands/ImportGreenlight3Command.php:
```diff
@@ -0,0 +1,429 @@
+                    hint: '(Optional).'
+                );
+
+            // Ask user what room type the imported rooms should get
```
Wrong comment

samuelwei replied:
fixed. @pizkaz Please check https://github.com/THM-Health/PILOS/pull/2665/changes/f79016d28aa69df3a25964201d347658b21734a7 if i didn't come to a false conclusion

pizkaz replied:
You're right. See [code](https://github.com/THM-Health/PILOS/pull/2665/changes/BASE..fe328c6106b27a22fe922fb76ea0b590926e9d9c#diff-f778303335f959e289e74d7bd691ea1a7932a223720c8134d41e9a378388d981R191-R193)

samuelwei committed:
Code cleanup

samuelwei committed:
Fix comment in code and docs on default role

samuelwei committed:
Docs: Remove false .env reference for GL3 import

samuelwei committed:
Fix tests

samuelwei commented on app/Console/Commands/ImportGreenlight3Command.php:
```diff
@@ -0,0 +1,429 @@
+use App\Settings\GeneralSettings;
+use Config;
+use DB;
+use Hash;
```
Avoid global namespace for laravel facades

samuelwei replied:
fixed

samuelwei committed:
Code style

Avoid global namespace for laravel facades

samuelwei commented on app/Http/Controllers/api/v1/RoomController.php:
```diff
@@ -467,7 +467,7 @@ public function authenticate(Room $room, RoomAuthRequest $request)
             $accessCode = $request->access_code;
 
-            if (is_numeric($accessCode) && $room->access_code == $accessCode) {
+            if ($room->access_code == $accessCode) {
```
the database column is string, see https://github.com/THM-Health/PILOS/blob/develop/database/migrations/2025_09_01_153901_change_access_code_to_string.php

samuelwei committed:
Fix missing changelog link

pizkaz commented:
~~I fear the Code Rabbit AI made me break room settings. Investigating...~~

It was the room export mode issue you found and fixed.

samuelwei committed:
Use strict equal checking for access code

samuelwei committed:
Fix test bug

samuelwei committed:
Improve docs

samuelwei committed:
Fix php docblock

samuelwei committed:
Fix missing expert mode for imported rooms

# Conflicts:
#	CHANGELOG.md

samuelwei committed:
Add join route for GL3 compatibility mode

pizkaz committed:
chore: Rename Greenlight import test helpers

Clarifies which helper is used where. GreenlightPresentation will be
re-used for GL2 presentation import.

pizkaz committed:
impr: Clean up GL3 import command test

pizkaz committed:
doc: Add missing import:greenlight-v3 command option

pizkaz committed:
impr: Explicitly filter room presentations from GL3 active storage attachments
